File: app/src/main/java/com/cliniva/enventory/utils/string/StringSpanBuilder.java
package com.cliniva.enventory.utils.string;

import android.graphics.Typeface;
import android.text.SpannableStringBuilder;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

public class StringSpanBuilder {

    private static List<SpanSection> spanSections;
    private static StringBuilder stringBuilder;

    private StringSpanBuilder() {
        // Disable class to not publicly instantiable
    }

    public static StringSpanBuilder getInstance() {

        stringBuilder = new StringBuilder();
        spanSections = new ArrayList<>();

        return new StringSpanBuilder();
    }

    public StringSpanBuilder append(String text) {
        spanSections.add(SpanSection.getInstance(text).setStartIndex(stringBuilder.length()));
        stringBuilder.append(text);
        return this;
    }

    public StringSpanBuilder append(SpanSection spanSection) {

        spanSections.add(spanSection.setStartIndex(stringBuilder.length()));
        stringBuilder.append(spanSection.getText());
        return this;
    }

    public StringSpanBuilder appendNewLine(){
        stringBuilder.append("\n");
        return this;
    }

    public void buildWithTextView(TextView textView) {
        textView.setText(build(), TextView.BufferType.SPANNABLE);
    }

    private SpannableStringBuilder build() {

        SpannableStringBuilder spannableStringBuilder = new SpannableStringBuilder(stringBuilder.toString());

        for (SpanSection section : spanSections) {
            section.apply(spannableStringBuilder);
        }

        return spannableStringBuilder;
    }
}

File: app/src/main/java/com/amirsons/inventory/utils/stringspan/StringSpanBuilder.kt
package com.amirsons.inventory.utils.stringspan

import android.text.SpannableStringBuilder
import android.widget.TextView

// Disable class to not publicly instantiable
class StringSpanBuilder private constructor() {

    private var spanSections: ArrayList<SpanSection> = ArrayList()
    private var stringBuilder: StringBuilder = StringBuilder()

    companion object {
        @JvmStatic
        val instance: StringSpanBuilder
            get() = StringSpanBuilder()
    }

    fun append(text: String): StringSpanBuilder {
        spanSections.add(SpanSection.getInstance(text).setStartIndex(stringBuilder.length))
        stringBuilder.append(text)
        return this
    }

    fun append(spanSection: SpanSection): StringSpanBuilder {
        spanSections.add(spanSection.setStartIndex(stringBuilder.length))
        stringBuilder.append(spanSection.spanText)
        return this
    }

    fun appendNewLine(): StringSpanBuilder {
        stringBuilder.append("\n")
        return this
    }

    fun buildWithTextView(textView: TextView) {
        textView.setText(build(), TextView.BufferType.SPANNABLE)
    }

    private fun build(): SpannableStringBuilder {

        val spannableStringBuilder = SpannableStringBuilder(stringBuilder.toString())

        for (section in spanSections) {
            section.apply(spannableStringBuilder)
        }

        return spannableStringBuilder
    }
}